wxMSW: Fix for dark drop down icon in dark mode

sz5000 opened the pull request:
As you can see here, the dropdown arrows are barely visible and this is an MSW issue as there is no way to change that:

![image](https://github.com/user-attachments/assets/212b96a5-8213-4f18-af1b-dc97b8c4b252)

Drawing of the icon is now customized and uses the foreground color now:

![image](https://github.com/user-attachments/assets/c490f648-1e32-45c1-a0b2-6e72de3d9b83)

sz5000 committed:
fix for dark drop down icon in dark mode

Drawing of the icon is customized and uses the foreground color now.

vadz commented on src/msw/toolbar.cpp:
```diff
@@ -1694,8 +1694,50 @@ bool wxToolBar::MSWOnNotify(int WXUNUSED(idCtrl),
+
+            case CDDS_ITEMPOSTPAINT: {
+                // custom draw the drop-down arrow here, as it is always black
+                TBBUTTONINFO bi = { 0 };
+                bi.cbSize = sizeof(TBBUTTONINFO);
```
For consistency please use
```suggestion
                WinStruct<TBBUTTONINFO> bi;
```

vadz commented on src/msw/toolbar.cpp:
```diff
@@ -1694,8 +1694,50 @@ bool wxToolBar::MSWOnNotify(int WXUNUSED(idCtrl),
+                    ::SendMessage(GetHwnd(), TB_GETITEMDROPDOWNRECT, (WPARAM)itemIndex, (LPARAM)&ddrc);
+
+                    if (nmtbcd->nmcd.uItemState & CDIS_HOT) {
+                        HBRUSH bgBrush = CreateSolidBrush(wxColourToRGB(GetBackgroundColour().ChangeLightness(110)));
```
We have RAII helper for brushes:
```suggestion
                        AutoHBRUSH bgBrush(wxColourToRGB(GetBackgroundColour().ChangeLightness(110)));
```

vadz commented on src/msw/toolbar.cpp:
```diff
@@ -1694,8 +1694,50 @@ bool wxToolBar::MSWOnNotify(int WXUNUSED(idCtrl),
+                        ::FillRect(nmtbcd->nmcd.hdc, &ddrc, bgBrush);
+                        DeleteObject(bgBrush);
+                    }
+                    else {
+                        HBRUSH bgBrush = CreateSolidBrush(wxColourToRGB(GetBackgroundColour()));
+                        ::FillRect(nmtbcd->nmcd.hdc, &ddrc, bgBrush);
+                        DeleteObject(bgBrush);
+                    }
```
It looks like we should merge with the `if` branch and just use a different colour here.

sz5000 replied:
No, not just the else case, otherwise you won't see any separation between the dropdown and the actual button when you hover over it with the mouse.

vadz commented on src/msw/toolbar.cpp:
```diff
@@ -1694,8 +1694,50 @@ bool wxToolBar::MSWOnNotify(int WXUNUSED(idCtrl),
+                        { arrowCenterX, arrowCenterY + FromDIP(2) }
+                    };
+
+                    HBRUSH fgBrush = CreateSolidBrush(wxColourToRGB(GetForegroundColour()));
```
```suggestion
                    AutoHBRUSH fgBrush(wxColourToRGB(GetForegroundColour()));
```

vadz commented on src/msw/toolbar.cpp:
```diff
@@ -1694,8 +1694,50 @@ bool wxToolBar::MSWOnNotify(int WXUNUSED(idCtrl),
+                    };
+
+                    HBRUSH fgBrush = CreateSolidBrush(wxColourToRGB(GetForegroundColour()));
+                    HPEN hPen = CreatePen(PS_SOLID, 1, wxColourToRGB(GetForegroundColour()));
```
```suggestion
                    AutoHPEN hPen(wxColourToRGB(GetForegroundColour()));
```

vadz reviewed (commented):
Thanks, I do see this problem in the toolbar sample, but applying this results in background colour being different from the rest of the toolbar:

![image](https://github.com/user-attachments/assets/9769a52f-90ed-4182-adb3-b33f952e5724)

Is it different under Windows 11?

sz5000 committed:
Update src/msw/toolbar.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

sz5000 committed:
Update src/msw/toolbar.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

sz5000 committed:
Update src/msw/toolbar.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

sz5000 committed:
Update src/msw/toolbar.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

sz5000 committed:
nmtbcd->clrHighlightHotTrack changed as if toolbar background color is same as wxSysColourToRGB(wxSYS_COLOUR_HOTLIGHT), we would have no visual effect when hovering.

Background color is determined via m_hasBgCol ? GetBackgroundColour() : wxSystemSettings::GetColour(wxSYS_COLOUR_BTNFACE) because this also used in wxToolBar::MSWGetToolbarBgBrush() and MSWGetToolbarBgBrush is used in erase event. If we would just use GetBackgroundColour() in CDDS_ITEMPOSTPAINT phase, then the dropdown background and toolbar background are different if the toolbar background is not customized.

sz5000 commented:
I changed the code further and I have tested with the sample and different background colors and I think the current code fixes all issues.

`clrHighlightHotTrack` changed as if toolbar background color is the same as `wxSysColourToRGB(wxSYS_COLOUR_HOTLIGHT)`, we would have no visual effect when hovering.

Background color is determined via `m_hasBgCol ? GetBackgroundColour() : wxSystemSettings::GetColour(wxSYS_COLOUR_BTNFACE)` because this also used in `wxToolBar::MSWGetToolbarBgBrush()` and `MSWGetToolbarBgBrush` is used in erase event. If we would just use `GetBackgroundColour()` in `CDDS_ITEMPOSTPAINT` phase, then the dropdown background and toolbar background are different if the toolbar background is not customized.

Lightness of dropdown area and actual toolbar button is a little different when hovering in order to make it clear to the user, that the drop-down area has different behavior then the actual button. Otherwise the user would not recognize where the drop-down area is.

For me, this is OK now. Please have a final look.

sz5000 committed:
Code style changes

vadz commented:
Sorry, I'm confused by the conflicts between this one and #24900: that PR added custom background drawing and `TBCDRF_NOBACKGROUND` while this one sets `clrHighlightHotTrack`. Should we do both of them, i.e. set `clrHighlightHotTrack` but still draw background ourselves in the non-hot case?

sz5000 commented:
`clrHighlightHotTrack` is used when the mouse hovers over a toolbar button.

We only draw background ourselves if item state is CDIS_CHECKED but not CDIS_HOT.

The code should finally be:

```
case CDDS_ITEMPREPAINT:
{
    // If we get here, we must have returned CDRF_NOTIFYITEMDRAW
    // from above, so we're using the dark mode and need to
    // customize the colours for it.
    nmtbcd->clrText =
        nmtbcd->clrTextHighlight = wxColourToRGB(GetForegroundColour());

    wxColour colBg = m_hasBgCol ? GetBackgroundColour() : wxSystemSettings::GetColour(wxSYS_COLOUR_BTNFACE);
    nmtbcd->clrHighlightHotTrack = wxColourToRGB(colBg.ChangeLightness(115));

    // draw custom checked button background
    if ( (nmtbcd->nmcd.uItemState & (CDIS_CHECKED | CDIS_HOT)) == CDIS_CHECKED )
    {
        const wxColor color = wxSystemSettings::GetColour(wxSYS_COLOUR_HOTLIGHT).ChangeLightness(110);
        AutoHBRUSH br(wxColourToRGB(color));
        ::FillRect(nmtbcd->nmcd.hdc, &nmtbcd->nmcd.rc, br);
        *result = CDRF_DODEFAULT | TBCDRF_USECDCOLORS | TBCDRF_HILITEHOTTRACK | TBCDRF_NOBACKGROUND | CDRF_NOTIFYPOSTPAINT;
        return true;
    }

    *result = CDRF_DODEFAULT | TBCDRF_USECDCOLORS | TBCDRF_HILITEHOTTRACK | CDRF_NOTIFYPOSTPAINT;
    return true;
}

case CDDS_ITEMPOSTPAINT:
{
    // custom draw the drop-down arrow here, as it is always black
    WinStruct<TBBUTTONINFO> bi;
    bi.dwMask = TBIF_STYLE | TBIF_COMMAND;
    int itemIndex = (int) ::SendMessage(GetHwnd(), TB_GETBUTTONINFO, (WPARAM)nmtbcd->nmcd.dwItemSpec, (LPARAM)&bi);
    if (itemIndex >= 0 && bi.fsStyle & TBSTYLE_DROPDOWN)
    {
        RECT ddrc = { 0 };
        ::SendMessage(GetHwnd(), TB_GETITEMDROPDOWNRECT, (WPARAM)itemIndex, (LPARAM)&ddrc);

        wxColour colBg = m_hasBgCol ? GetBackgroundColour() : wxSystemSettings::GetColour(wxSYS_COLOUR_BTNFACE);
        if (nmtbcd->nmcd.uItemState & CDIS_HOT)
        {
            AutoHBRUSH bgBrush(wxColourToRGB(colBg.ChangeLightness(120)));
            ::FillRect(nmtbcd->nmcd.hdc, &ddrc, bgBrush);
        }
        else
        {
            AutoHBRUSH bgBrush(wxColourToRGB(colBg));
            ::FillRect(nmtbcd->nmcd.hdc, &ddrc, bgBrush);
        }

        int arrowCenterX = (ddrc.left + ddrc.right) / 2;
        int arrowCenterY = (ddrc.top + ddrc.bottom) / 2;
        POINT ptsArrow[3] =
        {
            { arrowCenterX - FromDIP(3), arrowCenterY - FromDIP(2)},
            { arrowCenterX + FromDIP(3), arrowCenterY - FromDIP(2) },
            { arrowCenterX, arrowCenterY + FromDIP(2) }
        };

        AutoHBRUSH fgBrush(wxColourToRGB(GetForegroundColour()));
        AutoHPEN hPen(wxColourToRGB(GetForegroundColour()));
        SelectObject(nmtbcd->nmcd.hdc, hPen);
        SelectObject(nmtbcd->nmcd.hdc, fgBrush);
        Polygon(nmtbcd->nmcd.hdc, ptsArrow, 3);
    }

    *result = CDRF_DODEFAULT;
    return true;
}
```

vadz commented:
Looks nice, thanks! I've merged both PRs now with only some minor style changes, please let me know if I did anything wrong.

Thanks again!

sz5000 commented:
Thanks, looks good so far.
